Fix to asynchttpserver form data/body broken with #13147

mrhdias opened the pull request:
This fix add a new server option that allow the request body to be processed outside the asynchttpserver library to break big files into chunks of data. This change does not break anything and works with the jester and the rosencrantz.

The new server option "stream" when it is true uses the Future Streams otherwise all body is readed at once. The "stream" option by default is set to false.

The idea of this option I took from the python requests module (https://requests.readthedocs.io/en/v1.2.3/api/).
"When stream=True is set on the request, this avoids reading the content at once into memory for large responses. The chunk size is the number of bytes it should read into memory. This is not necessarily the length of each item returned as decoding can take place."

Example:
```nim
let server = newAsyncHttpServer(stream = true) # if true use Future Streams
waitFor server.serve(Port(8080), cb)
```

mrhdias committed:
Fix to asynchttpserver form data/body broken with #13147

This fix add a new server option that allow the request body to be processed outside the asynchttpserver library to break big files into chunks of data. This change does not break anything and works with the jester and the rosencrantz.

The new server option "stream" when it is true uses the Future Streams otherwise all body is readed at once. The "stream" option by default is set to false.

The idea of this option I took from the python requests module (https://requests.readthedocs.io/en/v1.2.3/api/).
"When stream=True is set on the request, this avoids reading the content at once into memory for large responses. The chunk size is the number of bytes it should read into memory. This is not necessarily the length of each item returned as decoding can take place."

Example:
let server = newAsyncHttpServer(stream = true) # if true use Future Streams
waitFor server.serve(Port(8080), cb)

mrhdias committed:
Fixed a small text error!

Araq commented on lib/pure/asynchttpserver.nim:
```diff
@@ -91,46 +91,34 @@ export httpcore except parseHeader
+    url*: Uri
+    hostname*: string    ## The hostname of the client that made the request.
+    body*: string
+    bodyStream*: FutureStream[string]
```
Make it a private field please and add accessor procs for NimVersion >= (1, 1).

dom96 replied:
@Araq what is the purpose of these `NimVersion >= (1,1)` annotations? Are you intending to reuse the newest stdlib for older versions of Nim?

Araq replied:
It's so that when we release Nim 1.2 it can compile every 1.0 project via ``--useVersion:1.0``. Most code won't need --useVersion, but it's for an extra layer of safety since we want to maintain v1.0 for a couple of years.

dom96 replied:
> Most code won't need --useVersion, but it's for an extra layer of safety since we want to maintain v1.0 for a couple of years.

I'm going to challenge you here, because I think this is a bad idea. This code is already becoming really convoluted: we're duplicating code with small changes here and there. Who is going to maintain each branch? What is the point of this at all?

If you want to compile a project as if it's v1.0 then why not just use Nim 1.0?

Araq replied:
> Who is going to maintain each branch?

Well certainly not you anyway.

> What is the point of this at all?
> If you want to compile a project as if it's v1.0 then why not just use Nim 1.0?

Because we cannot backport bugfixes into 1.0 forever and the competitors have flags like ``-std=c++11`` and it works.

dom96 commented on lib/pure/asynchttpserver.nim:
```diff
@@ -91,46 +91,34 @@ export httpcore except parseHeader
     reuseAddr: bool
     reusePort: bool
     maxBody: int ## The maximum content-length that will be read for the body.
+    stream: bool ## By default (stream = false), the body of the request is readed immediately
```
```suggestion
    stream: bool ## By default (stream = false), the body of the request is read immediately
```

dom96 commented on lib/pure/asynchttpserver.nim:
```diff
@@ -315,7 +300,7 @@ proc processRequest(
+      if server.stream:
         var remainder = contentLength
         while remainder > 0:
           let readSize = min(remainder, chunkSize)
```
I might be missing something, but this whole streaming code looks wrong to me. You are going to be writing all the data into the bodyStream so you are effectively just reimplementing what happens when `stream` isn't set. How does this enable streaming when the code which uses the `Request` only gets it after the `bodyStream` is written to fully?

mrhdias replied:
You are right -( please check the new version that use a iterator to read the chunks of data.

dom96 reviewed (changes requested):
It seems to me like #13147 shouldn't have been merged. See my comment.

mrhdias committed:
New implementation that use a interator instance of future streams

The previous suggestion don't work with future streams because all data is write to the the future stream and used later, but only when the queue is full and the future is complete as Dominik Picheta mentioned in the comment. This new version uses an iterator to read the chunks of data.

mrhdias commented:
> This fix add a new server option that allow the request body to be processed outside the asynchttpserver library to break big files into chunks of data. This change does not break anything and works with the jester and the rosencrantz.
> 
> The new server option "stream" when it is true uses the Future Streams otherwise all body is readed at once. The "stream" option by default is set to false.
> 
> The idea of this option I took from the python requests module (https://requests.readthedocs.io/en/v1.2.3/api/).
> "When stream=True is set on the request, this avoids reading the content at once into memory for large responses. The chunk size is the number of bytes it should read into memory. This is not necessarily the length of each item returned as decoding can take place."
> 
> Example:
> 
> ```nim
> let server = newAsyncHttpServer(stream = true) # if true use Future Streams
> waitFor server.serve(Port(8080), cb)
> ```

The previous suggestion with future streams may not work as expected because all data is write to the future stream and used later, but only when the queue is full and the future is complete as Dominik Picheta mentioned in the comment.

This new version uses an iterator to read the chunks of data and works fine with jester and the rosencrantz without any modification.

Example with iterator
```nim
import asynchttpserver, asyncdispatch
import strutils, strformat

proc htmlpage(contentLength, bodyLength: int): string =
  return &"""
<!Doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8"/>
  </head>
  <body>
    <form action="/" method="post" enctype="multipart/form-data">
      File: <input type="file" name="testfile" accept="text/*"><br />
      <input style="margin:10px 0;" type="submit">
    </form><br />
    Expected Body Length: {contentLength} bytes<br />
    Actual Body Length: {bodyLength} bytes
  </body>
</html>
"""

proc cb(req: Request) {.async.} =
  var
    contentLength = 0
    bodyLength = 0
  if req.reqMethod == HttpPost:
    contentLength = req.headers["Content-length"].parseInt
    # Read 8*1024 bytes at a time
    # optional chunkSize parameter. The default is 8*1024
    for length, data in req.bodyStream(8*1024):
      let content = await data
      if length == content.len:
        bodyLength += content.len
      else:
        # Handle exception
        await req.respond(Http400,
          "Bad Request. Data read has a different length than the expected.")
        return
  await req.respond(Http200, htmlpage(contentLength, bodyLength))

let server = newAsyncHttpServer(maxBody = 10485760, stream = true) # 10 MB
waitFor server.serve(Port(8080), cb)
```

Araq commented on lib/pure/asynchttpserver.nim:
```diff
@@ -65,14 +68,55 @@
+##    proc htmlpage(contentLength, bodyLength: int): string =
+##      return &"""
+##    <!Doctype html>
+##    <html lang="en">
```
Excessive example, can we shorten it somehow?

mrhdias replied:
I already removed one of the examples.

mrhdias committed:
Two examples have been condensed into one.

The two examples of how to handle the post request have been condensed into one.

mrhdias committed:
Fix a small error!

mrhdias committed:
asynchttpserver now can handle chunks of data.

Added an iterator to asynchttpserver that allows the request body to be read in chunks of data when new server "stream" option is set to true.

dom96 commented on lib/pure/asynchttpserver.nim:
```diff
@@ -193,13 +214,30 @@ proc parseProtocol(protocol: string): tuple[orig: string, major, minor: int] =
+    ##
+    ## This iterator return a tuple with the length of the data that was read
+    ## and a future.
+    var remainder = request.contentLength
+    while remainder > 0:
+      let readSize = min(remainder, chunkSize)
+      let data = request.client.recv(readSize)
+      if data.failed:
+        raise newException(ValueError, "Error reading POST data from client.")
+      yield (readSize, data)
+      remainder -= readSize
```
This iterator is quite broken and makes me wonder whether you've tested this at all:

* `recv` reads **up to** `readSize`, it doesn't guarantee you'll get that much data, and a lot of the time you won't.

* you're checking whether the `data` future failed, without `await`ing it. This code path will almost never trigger.

mrhdias replied:
You can test with the example and you will see that it works. Try uploading a 900 MB file! Please give me a solution or a way to implement this.

mrhdias replied:
> This iterator is quite broken and makes me wonder whether you've tested this at all:
> 
>     * `recv` reads **up to** `readSize`, it doesn't guarantee you'll get that much data, and a lot of the time you won't.
> 
>     * you're checking whether the `data` future failed, without `await`ing it. This code path will almost never trigger.
From the nim documentation (https://nim-lang.org/0.16.0/asyncdispatch.html)
```nim
var future = sock.recv(100)
yield future
if future.failed:
  # Handle exception
```
I can't use the await in the iterators!
https://github.com/nim-lang/Nim/issues/3885
Async iterators #6576
https://github.com/nim-lang/Nim/issues/6576

dom96 reviewed (commented):
Thank you for working on this, but it shouldn't have been merged. I cannot see how this can work. Iterators returning futures seems like an anti-pattern in general to me.

dom96 commented:
@mrhdias @Araq is this going to be fixed? If not I'll be reverting it.

Araq commented:
Looks like you have to revert it as @mrhdias is unresponsive.

dom96 commented:
I created a PR.

mrhdias commented:
> Looks like you have to revert it as @mrhdias is unresponsive.

Is it possible to improve the code to read chunks of data? this is important, now it is impossible read big files without exhausting all resources!

Araq commented:
Well better than reverting it is to fix it, of course.

mrhdias commented:
> Well better than reverting it is to fix it, of course.

Is possible make something like this?
Asynchronous Iteration Support
Stream reader supports asynchronous iteration over BODY.
https://docs.aiohttp.org/en/stable/streams.html

ITwrx commented:
@Araq can someone help @mrhdias get this worked out? he's trying to find a way to fix this, but needs some help. I don't know enough to help. It seems to me (maybe i'm wrong), that until this issue is resolved, the whole nim web ecosystem can't upload files in a production-acceptable way. That's a big problem for my current majority use case for nim.

Araq commented:
I'm willing to help but first somebody needs to lecture me why we don't simply use [sendfile](https://man7.org/linux/man-pages/man2/sendfile.2.html) ?

ITwrx commented:
I don't know for sure if you mean `in asynchttpserver` or `by the end user`, but if the former i'm currently too "high level" to discuss it intelligently, much less lecture. :) Or did you mean the latter? I've never used sendfile directly that i remember. 

I'm just a new nim end user that needs to be able to upload files via http without using too much ram and/or not releasing the ram. Preferably via a well thought out web framework, with documentation that explains it to me like i'm five years old. :) 

I'll submit a commercial support quote request, as i'm sure you have plenty of things you could be working on.

dom96 commented:
We totally should use `sendfile` :)

ITwrx commented:
@dom96 thanks for the PR link.  

I didn't know this was all being discussed already. I could have looked around more.
